[SPARK-22000][SQL] Address missing Upcast in JavaTypeInference.deserializerFor

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

Spark expects the type of column and the type of matching field is same when deserializing to Object, but Spark hasn't actually restrict it (at least for Java bean encoder) and some users just do it and experience undefined behavior (in SPARK-22000, Spark throws compilation failure on generated code because it calls `.toString()` against primitive type.

It doesn't produce error in Scala side because `ScalaReflection.deserializerFor` properly inject Upcast if necessary. This patch proposes applying same thing to `JavaTypeInference.deserializerFor` as well.

Credit to @srowen, @maropu, and @cloud-fan since they provided various approaches to solve this.

## How was this patch tested?

Added UT which query is slightly modified based on sample code in attachment on JIRA issue.

HeartSaVioR committed:
[SPARK-22000][SQL] Use String.valueOf to assign value to String type of field in Java Bean

HeartSaVioR commented:
cc. @maropu @kiszk @srowen since they had discussion on JIRA issue.

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
ok to test

maropu commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -115,6 +115,37 @@ public void testBeanWithMapFieldsDeserialization() {
+      .option("header", "true")
+      .option("mode", "DROPMALFORMED")
+      .schema("ref int, userId string, x int, y int")
+      .load("src/test/resources/test-data/spark-22000.csv")
```
We need to read test data from a file instead of `spark.createDataFrame(...`?

HeartSaVioR replied:
I guess not. Let me try to change not to use file.

maropu replied:
Thanks!

maropu commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -252,4 +283,73 @@ public String toString() {
+    @Override
+    public String toString() {
+      return String.format("ref='%s', userId='%s', x=%d, y=%d", ref, userId, x, y);
+    }
```
We need to override `toString`, `hashCode`, and `equals` for the test?

HeartSaVioR replied:
It would be easier to test with them, otherwise we need to add code for comparison anyway. Other existing tests also leveraged them.

maropu commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -115,6 +115,37 @@ public void testBeanWithMapFieldsDeserialization() {
+      .format("csv")
+      .option("header", "true")
+      .option("mode", "DROPMALFORMED")
+      .schema("ref int, userId string, x int, y int")
```
Can you add tests for more types other than `int`?

HeartSaVioR committed:
Address review comments from maropu

HeartSaVioR commented:
Thanks @maropu I addressed your comments. The code got much longer - in Java world I had to deal with Row directly, as well as I need to add equals() / hashCode() (this is actually optional but best practice would implement along with equals()) / toString(). Hope this is OK.

srowen commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -115,6 +124,70 @@ public void testBeanWithMapFieldsDeserialization() {
     Assert.assertEquals(records, MAP_RECORDS);
   }
 
+  private static final List<Row> ROWS_SPARK_22000 = new ArrayList<>();
```
All the static initialization could just be in a static block rather than split it up, but could this all be local to the new test method if that's the only place it's used?

HeartSaVioR replied:
I just followed the approach the test class already has, but I agree they can be local to new test method. Will address.

srowen commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -252,4 +325,116 @@ public String toString() {
     }
   }
+
+  public static class RecordSpark22000 {
```
`final` just to be tidy?

srowen commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -252,4 +325,116 @@ public String toString() {
+    }
+
+    @Override
+    public String toString() {
```
Does this need toString? I understand hashCode and equals

HeartSaVioR replied:
It will help to compare expected and actual when test fails. Otherwise they would've seen as `Object.toString()` does and it doesn't provide any information why they are not equal.

srowen reviewed (changes requested):
I put some minor comments on the test, but, I am not sure this change matches the JIRA? The issue concerned calling `.toString` on primitive types, not `String`. I suggested `String.valueOf` because it accepts primitives.

This change alters behavior a little bit. Now a null is deserialized as `"null"` not `null`. I'm not sure we want that, nor was that an issue before; can this be called on null anyway?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #102584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102584/testReport)** for PR 23854 at commit [`443e74f`](https://github.com/apache/spark/commit/443e74f207d4869d11944f4d95df97f5bc40e6bc).

SparkQA commented:
**[Test build #102584 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102584/testReport)** for PR 23854 at commit [`443e74f`](https://github.com/apache/spark/commit/443e74f207d4869d11944f4d95df97f5bc40e6bc).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102584/
Test FAILed.

HeartSaVioR commented:
First of all, I'm not a catalyst expert so I might be wrong, so please correct me if I'm mistaken.

> The issue concerned calling `.toString` on primitive types, not `String`.

Why it was happening is IntegerType is matched to `String type of field` when deserializing to Java Bean. If it was matched to Integer type of field, `Integer.valueOf()` would be called.

Looks like Spark doesn't just restrict matching type should be same. It handles some implicit type conversion via leveraging Java API, and primitive type was not compatible with String (case it calls `.toString()`) whereas it can be matched in other way (`String.valueOf()`).

> This change alters behavior a little bit. Now a null is deserialized as `"null"` not `null`. I'm not sure we want that, nor was that an issue before; can this be called on null anyway?

Unless I misread javadoc or javadoc is wrong, `StaticInvoke` will take care of that. This change would not incur calling `String.valueOf(null)`, otherwise passing null should be happening to other types in same if statement as well and it will just throw NumberFormatException for numeric types.

https://github.com/apache/spark/blob/17d0cfcaa4a43fd55b81065d907538a9c1bf569b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala#L199-L220

Btw I guess the case can be added in new UT - we don't need to worry about it after we apply the case in UT. I'll try.

HeartSaVioR committed:
Address review comments from srowen, as well as fix style

SparkQA commented:
**[Test build #102598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102598/testReport)** for PR 23854 at commit [`d869bba`](https://github.com/apache/spark/commit/d869bbafe9c499c09b994331f280d53cc52fceb9).

srowen commented:
I agree with your argument about null; it would never have worked here anyway. It feels like there should be an assertion about it to make sure, as if a null reaches here it returns "null" rather than an exception.

I also don't know this part well enough, but primitives are matched to string? That sounds weird. It seems like primitives are already handled in 
```
      case c if !inferExternalType(c).isInstanceOf[ObjectType] => path
```

It wasn't clear from the JIRA, initially, how this even happened. Now there's a simple test case attached. Could we paste that in as a test here, and see if it even fails without the change? and if so, how.

The test case suggests that a String field is the problem, but it has toString and isn't handled by this part.

HeartSaVioR commented:
> Now there's a simple test case attached. Could we paste that in as a test here, and see if it even fails without the change? and if so, how.

The added UT was slightly modified from attached test case: I just modified it based on review comments. Previously added test failed without the change, and the revised test remains the same.

> I agree with your argument about null; it would never have worked here anyway. It feels like there should be an assertion about it to make sure, as if a null reaches here it returns "null" rather than an exception.

No I meant null has been handled well and modified code also works well with null. Please find `String.valueOf` from generated code - it is guarded with checking null. If the value is null, `String.valueOf` is not called and the result is null. That's what `propagateNull` does for us.

```
07:13:10.684 DEBUG org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection: code for initializejavabean(newInstance(class test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000), (setBooleanField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[6, boolean, true], true, false)), (setLongField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[2, bigint, true], true, false)), (setIntField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[1, int, true], true, false)), (setStringField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[5, string, true], true, false)), (setDoubleField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[4, double, true], true, false)), (setShortField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[0, smallint, true], true, false)), (setFloatField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[3, float, true], true, false)), (setTimestampField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[7, timestamp, true], true, false)), (setNullIntField,staticinvoke(class java.lang.String, ObjectType(class java.lang.String), valueOf, input[8, int, true], true, false))):
/* 001 */ public java.lang.Object generate(Object[] references) {
/* 002 */   return new SpecificSafeProjection(references);
/* 003 */ }
/* 004 */
/* 005 */ class SpecificSafeProjection extends org.apache.spark.sql.catalyst.expressions.codegen.BaseProjection {
/* 006 */
/* 007 */   private Object[] references;
/* 008 */   private InternalRow mutableRow;
/* 009 */   private boolean resultIsNull_0;
/* 010 */   private boolean argValue_0;
/* 011 */   private boolean resultIsNull_1;
/* 012 */   private long argValue_1;
/* 013 */   private boolean resultIsNull_2;
/* 014 */   private int argValue_2;
/* 015 */   private boolean resultIsNull_3;
/* 016 */   private boolean resultIsNull_4;
/* 017 */   private double argValue_3;
/* 018 */   private boolean resultIsNull_5;
/* 019 */   private short argValue_4;
/* 020 */   private boolean resultIsNull_6;
/* 021 */   private float argValue_5;
/* 022 */   private boolean resultIsNull_7;
/* 023 */   private long argValue_6;
/* 024 */   private boolean resultIsNull_8;
/* 025 */   private int argValue_7;
/* 026 */   private UTF8String[] mutableStateArray_0 = new UTF8String[1];
/* 027 */
/* 028 */   public SpecificSafeProjection(Object[] references) {
/* 029 */     this.references = references;
/* 030 */     mutableRow = (InternalRow) references[references.length - 1];
/* 031 */
/* 032 */
/* 033 */   }
/* 034 */
/* 035 */   public void initialize(int partitionIndex) {
/* 036 */
/* 037 */   }
/* 038 */
/* 039 */   public java.lang.Object apply(java.lang.Object _i) {
/* 040 */     InternalRow i = (InternalRow) _i;
/* 041 */     final test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000 value_1 = false ?
/* 042 */     null : new test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000();
/* 043 */     test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000 javaBean_0 = value_1;
/* 044 */     if (!false) {
/* 045 */       initializeJavaBean_0_0(i, javaBean_0);
/* 046 */       initializeJavaBean_0_1(i, javaBean_0);
/* 047 */       initializeJavaBean_0_2(i, javaBean_0);
/* 048 */     }
/* 049 */     if (false) {
/* 050 */       mutableRow.setNullAt(0);
/* 051 */     } else {
/* 052 */
/* 053 */       mutableRow.update(0, value_1);
/* 054 */     }
/* 055 */
/* 056 */     return mutableRow;
/* 057 */   }
/* 058 */
/* 059 */
/* 060 */   private void initializeJavaBean_0_2(InternalRow i, test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000 javaBean_0) {
/* 061 */
/* 062 */     resultIsNull_6 = false;
/* 063 */     if (!resultIsNull_6) {
/* 064 */       boolean isNull_15 = i.isNullAt(3);
/* 065 */       float value_15 = isNull_15 ?
/* 066 */       -1.0f : (i.getFloat(3));
/* 067 */       resultIsNull_6 = isNull_15;
/* 068 */       argValue_5 = value_15;
/* 069 */     }
/* 070 */
/* 071 */     boolean isNull_14 = resultIsNull_6;
/* 072 */     java.lang.String value_14 = null;
/* 073 */     if (!resultIsNull_6) {
/* 074 */       value_14 = java.lang.String.valueOf(argValue_5);
/* 075 */     }
/* 076 */     if (!isNull_14) {
/* 077 */       javaBean_0.setFloatField(value_14);
/* 078 */     }
/* 079 */
/* 080 */     resultIsNull_7 = false;
/* 081 */     if (!resultIsNull_7) {
/* 082 */       boolean isNull_17 = i.isNullAt(7);
/* 083 */       long value_17 = isNull_17 ?
/* 084 */       -1L : (i.getLong(7));
/* 085 */       resultIsNull_7 = isNull_17;
/* 086 */       argValue_6 = value_17;
/* 087 */     }
/* 088 */
/* 089 */     boolean isNull_16 = resultIsNull_7;
/* 090 */     java.lang.String value_16 = null;
/* 091 */     if (!resultIsNull_7) {
/* 092 */       value_16 = java.lang.String.valueOf(argValue_6);
/* 093 */     }
/* 094 */     if (!isNull_16) {
/* 095 */       javaBean_0.setTimestampField(value_16);
/* 096 */     }
/* 097 */
/* 098 */     resultIsNull_8 = false;
/* 099 */     if (!resultIsNull_8) {
/* 100 */       boolean isNull_19 = i.isNullAt(8);
/* 101 */       int value_19 = isNull_19 ?
/* 102 */       -1 : (i.getInt(8));
/* 103 */       resultIsNull_8 = isNull_19;
/* 104 */       argValue_7 = value_19;
/* 105 */     }
/* 106 */
/* 107 */     boolean isNull_18 = resultIsNull_8;
/* 108 */     java.lang.String value_18 = null;
/* 109 */     if (!resultIsNull_8) {
/* 110 */       value_18 = java.lang.String.valueOf(argValue_7);
/* 111 */     }
/* 112 */     if (!isNull_18) {
/* 113 */       javaBean_0.setNullIntField(value_18);
/* 114 */     }
/* 115 */
/* 116 */   }
/* 117 */
/* 118 */
/* 119 */   private void initializeJavaBean_0_1(InternalRow i, test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000 javaBean_0) {
/* 120 */
/* 121 */     resultIsNull_3 = false;
/* 122 */     if (!resultIsNull_3) {
/* 123 */       boolean isNull_9 = i.isNullAt(5);
/* 124 */       UTF8String value_9 = isNull_9 ?
/* 125 */       null : (i.getUTF8String(5));
/* 126 */       resultIsNull_3 = isNull_9;
/* 127 */       mutableStateArray_0[0] = value_9;
/* 128 */     }
/* 129 */
/* 130 */     boolean isNull_8 = resultIsNull_3;
/* 131 */     java.lang.String value_8 = null;
/* 132 */     if (!resultIsNull_3) {
/* 133 */       value_8 = java.lang.String.valueOf(mutableStateArray_0[0]);
/* 134 */     }
/* 135 */     if (!isNull_8) {
/* 136 */       javaBean_0.setStringField(value_8);
/* 137 */     }
/* 138 */
/* 139 */     resultIsNull_4 = false;
/* 140 */     if (!resultIsNull_4) {
/* 141 */       boolean isNull_11 = i.isNullAt(4);
/* 142 */       double value_11 = isNull_11 ?
/* 143 */       -1.0 : (i.getDouble(4));
/* 144 */       resultIsNull_4 = isNull_11;
/* 145 */       argValue_3 = value_11;
/* 146 */     }
/* 147 */
/* 148 */     boolean isNull_10 = resultIsNull_4;
/* 149 */     java.lang.String value_10 = null;
/* 150 */     if (!resultIsNull_4) {
/* 151 */       value_10 = java.lang.String.valueOf(argValue_3);
/* 152 */     }
/* 153 */     if (!isNull_10) {
/* 154 */       javaBean_0.setDoubleField(value_10);
/* 155 */     }
/* 156 */
/* 157 */     resultIsNull_5 = false;
/* 158 */     if (!resultIsNull_5) {
/* 159 */       boolean isNull_13 = i.isNullAt(0);
/* 160 */       short value_13 = isNull_13 ?
/* 161 */       (short)-1 : (i.getShort(0));
/* 162 */       resultIsNull_5 = isNull_13;
/* 163 */       argValue_4 = value_13;
/* 164 */     }
/* 165 */
/* 166 */     boolean isNull_12 = resultIsNull_5;
/* 167 */     java.lang.String value_12 = null;
/* 168 */     if (!resultIsNull_5) {
/* 169 */       value_12 = java.lang.String.valueOf(argValue_4);
/* 170 */     }
/* 171 */     if (!isNull_12) {
/* 172 */       javaBean_0.setShortField(value_12);
/* 173 */     }
/* 174 */
/* 175 */   }
/* 176 */
/* 177 */
/* 178 */   private void initializeJavaBean_0_0(InternalRow i, test.org.apache.spark.sql.JavaBeanDeserializationSuite$RecordSpark22000 javaBean_0) {
/* 179 */
/* 180 */     resultIsNull_0 = false;
/* 181 */     if (!resultIsNull_0) {
/* 182 */       boolean isNull_3 = i.isNullAt(6);
/* 183 */       boolean value_3 = isNull_3 ?
/* 184 */       false : (i.getBoolean(6));
/* 185 */       resultIsNull_0 = isNull_3;
/* 186 */       argValue_0 = value_3;
/* 187 */     }
/* 188 */
/* 189 */     boolean isNull_2 = resultIsNull_0;
/* 190 */     java.lang.String value_2 = null;
/* 191 */     if (!resultIsNull_0) {
/* 192 */       value_2 = java.lang.String.valueOf(argValue_0);
/* 193 */     }
/* 194 */     if (!isNull_2) {
/* 195 */       javaBean_0.setBooleanField(value_2);
/* 196 */     }
/* 197 */
/* 198 */     resultIsNull_1 = false;
/* 199 */     if (!resultIsNull_1) {
/* 200 */       boolean isNull_5 = i.isNullAt(2);
/* 201 */       long value_5 = isNull_5 ?
/* 202 */       -1L : (i.getLong(2));
/* 203 */       resultIsNull_1 = isNull_5;
/* 204 */       argValue_1 = value_5;
/* 205 */     }
/* 206 */
/* 207 */     boolean isNull_4 = resultIsNull_1;
/* 208 */     java.lang.String value_4 = null;
/* 209 */     if (!resultIsNull_1) {
/* 210 */       value_4 = java.lang.String.valueOf(argValue_1);
/* 211 */     }
/* 212 */     if (!isNull_4) {
/* 213 */       javaBean_0.setLongField(value_4);
/* 214 */     }
/* 215 */
/* 216 */     resultIsNull_2 = false;
/* 217 */     if (!resultIsNull_2) {
/* 218 */       boolean isNull_7 = i.isNullAt(1);
/* 219 */       int value_7 = isNull_7 ?
/* 220 */       -1 : (i.getInt(1));
/* 221 */       resultIsNull_2 = isNull_7;
/* 222 */       argValue_2 = value_7;
/* 223 */     }
/* 224 */
/* 225 */     boolean isNull_6 = resultIsNull_2;
/* 226 */     java.lang.String value_6 = null;
/* 227 */     if (!resultIsNull_2) {
/* 228 */       value_6 = java.lang.String.valueOf(argValue_2);
/* 229 */     }
/* 230 */     if (!isNull_6) {
/* 231 */       javaBean_0.setIntField(value_6);
/* 232 */     }
/* 233 */
/* 234 */   }
/* 235 */
/* 236 */ }
```

So it's completely safe to use `String.valueOf()` for String type, even other types in if statement.

HeartSaVioR commented:
```
      case c if !inferExternalType(c).isInstanceOf[ObjectType] => path
```

This would match when the type of field in Java class is primitive. `c` is not for type of column in Dataset but for Java class (and fields in class, with considering nested maybe?) we are creating Encoder. I guess you've got confusion here.

srowen commented:
OK agree about null.

I'm still trying to understand why the change fixes it, though I believe it does. Is it accidental or actually solving the problem?

Before your change, which field has the wrong generated code -- what type is it? the original test case suggests the String field is the problem, not the primitive problem. But then I don't get where the problem is coming from.

Just trying to understand if we're actually understanding why it happens?

HeartSaVioR commented:
> Before your change, which field has the wrong generated code -- what type is it?

The field/column was `ref`: the type from `ref` column in Dataset was `IntegerType` and the type of `ref` field in Java class was `String`.

To be honest I have no idea this is intended to support implicit type conversions when deserializing, but any types to String would sound convenient, and not much weird. Without knowing history/intention I guess I can't explain more.

HeartSaVioR commented:
@cloud-fan Could you help putting some knowledge around this issue, since you look to be author of Javabean Encoder? Thanks in advance!

HeartSaVioR commented:
Looks like Encoder is not aware of the type of columns in Dataset: then I understand why it doesn't have strict type check on columns and fields in Java bean. Still not sure what's the intention: expect only StringType for String field so safe to call `.toString()` there, or handle any types for String field so take `.toString()` to work with various types.

srowen reviewed (approved):
OK I'm getting this now. In the test case attached to the JIRA, the first column 'ref' is (correctly) inferred as IntegerType. The bean class has it as a String. That is arguably an error, but, it should be reasonable to make the assignment. Because it goes to a String, it is most correct to call `String.valueOf`, especially as that is consistent with how other types are handled and how the JDK API is set up, with `valueOf` methods for these boxed types and String. This should not change results, `String.valueOf(Object)` of course returns its `.toString`

SparkQA commented:
**[Test build #102598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102598/testReport)** for PR 23854 at commit [`d869bba`](https://github.com/apache/spark/commit/d869bbafe9c499c09b994331f280d53cc52fceb9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public final static class RecordSpark22000 `

maropu commented:
btw, why don't we cast internal data into compatible typed data before deserializing them into Bean instead of this weird type handling like int-> string in deserializer?

Probably, it seems we can add cast exprs before collecting data?
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L3360

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102598/
Test PASSed.

srowen commented:
That's probably a better design; this at least fixes the current design and bug. WDYT about proceeding anyway?

maropu commented:
Yea, sure. +1 for the fix in the current design. cc: @cloud-fan

HeartSaVioR commented:
Sure I think it sounds much better, but as a view of newcomer of Spark SQL it doesn't seem to be trivial.

> Probably, it seems we can add cast exprs before collecting data?
> https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L3360

I guess we also need to handle `Dataset.rdd` as well, and both path looks like going through `CatalystSerde.deserialize` to create DeserializeToObject. So I looked through how to deal with types of logical plan and deserializer, but `deserializer` has type of fields as JavaType (like `ObjectType(java.lang.String)` for String field) which we need to match and decide against Spark SQL type. Does Spark have some code to match, or even handle cast?

maropu commented:
We've already got SQL inferred types for java beans by `JavaTypeInference.inferDataType`, so I think its ok to just cast internal data into the inferred types somewhere.

HeartSaVioR commented:
The thing is, inferred types are not propagated into deserializer (which is `InitializeJavaBean`). It only has setters which `StaticInvoke` is passed for the String field, and it doesn't have information which type is required, only have information for return type of `StaticInvoke` which is wrapped with ObjectType.

So if we want to leverage inferred SQL DataType to avoid dealing with ObjectType, it may need to be also propagated in StaticInvoke (and also Invoke maybe) which also looks like non-trivial thing (the code change might be small but just not sure this is what we want).

HeartSaVioR commented:
Never mind. I realized `encoderFor[T]` has schema information. I'll try to apply the new approach.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -235,9 +236,6 @@ object JavaTypeInference {
           returnNullable = false)
 
-      case c if c == classOf[java.lang.String] =>
-        Invoke(path, "toString", ObjectType(classOf[String]))
```
The `ScalaReflection` does the same thing, do we have a problem there too?

AFAIK the `path` should be a string type column, and it's always safe to call `UTF8String.toString`. My gut feeling is, we miss to add `Upcast` somewhere in `JavaTypeInference`.

HeartSaVioR replied:
> AFAIK the `path` should be a string type column

The sample code tried to bind IntegerType column to String field in Java bean, which looks to break your expectation. (I guess ScalaReflection would not encounter this case.) 

Spark doesn't throw error for this case - actually Spark would show undefined behavior, compilation failures on generated code, even might be possible to throw runtime exception.

cloud-fan replied:
> The sample code tried to bind IntegerType column to String field in Java bean

In scala, we can also do this and Spark will add `Upcast`. e.g. `spark.range(1).as[String].collect` works fine.

I did a quick search and `JavaTypeInference` has no `Upcast`. We should fix it and follow `ScalaReflection`

HeartSaVioR replied:
Ah OK. I'll check and address it. Maybe it would be a separate PR if it doesn't fix the new test.

HeartSaVioR replied:
Yeah your suggestion seems to work nicely! I left comment to ask which approach to choose: please compare both approach and comment. Thanks!

SparkQA commented:
**[Test build #102656 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102656/testReport)** for PR 23854 at commit [`c4d611c`](https://github.com/apache/spark/commit/c4d611cd0f5b37dc5f6247ad9be7acc54730ad64).

HeartSaVioR commented:
@maropu Thanks for the idea. I applied your suggestion. 
@srowen I just rolled back the change of JavaTypeInference, and confirmed new test passed without this.

I couldn't inject Project(Exec) in `CatalystSerde.deserialize` since there's some case (`ExpressionEncoder.resolveAndBind()`) which doesn't use returned DeserializeToObject as LogicalPlan - it just leverages deserializer in DeserializeToObject which injected Project(Exec) is not effective.

SparkQA commented:
**[Test build #102656 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102656/testReport)** for PR 23854 at commit [`c4d611c`](https://github.com/apache/spark/commit/c4d611cd0f5b37dc5f6247ad9be7acc54730ad64).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102656/
Test FAILed.

HeartSaVioR commented:
It turns out to be more complicated than expected. I need to apply projection with cast only when deserialization happens, but some of cases I had to play with physical plan which I had to rebuild deserializer.

After the fix, nothing leverages `Dataset.deserializer` so I've removed it. According to comment, in sample case Spark should do type check and provide error, but as we've seen the error is not from type check so I'm wondering eagerly initializing `deserializer` does the type check properly.

SparkQA commented:
**[Test build #102665 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102665/testReport)** for PR 23854 at commit [`4063a37`](https://github.com/apache/spark/commit/4063a37e3a07123ae9cffde2dedb57cfea58f3ca).

SparkQA commented:
**[Test build #102665 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102665/testReport)** for PR 23854 at commit [`4063a37`](https://github.com/apache/spark/commit/4063a37e3a07123ae9cffde2dedb57cfea58f3ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102665/
Test FAILed.

SparkQA commented:
**[Test build #102679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102679/testReport)** for PR 23854 at commit [`661786b`](https://github.com/apache/spark/commit/661786b487f6a3a69e9c351d83a044f546100393).

HeartSaVioR committed:
Apply cloud-fan's suggestion: address Upcast in JavaTypeInference

HeartSaVioR commented:
Now I have another version of the path which is based on @cloud-fan's suggestion.
https://github.com/apache/spark/compare/master...HeartSaVioR:SPARK-22000-approach-cloud-fan

Please also compare with ScalaReflection when reading above change.

Please let me know which approach we prefer: I'll rebase this branch when we would prefer new approach, as well as update title and description on PR. Thanks in advance!

SparkQA commented:
**[Test build #102679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102679/testReport)** for PR 23854 at commit [`661786b`](https://github.com/apache/spark/commit/661786b487f6a3a69e9c351d83a044f546100393).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102679/
Test PASSed.

SparkQA commented:
**[Test build #102692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102692/testReport)** for PR 23854 at commit [`60ca088`](https://github.com/apache/spark/commit/60ca0889f14db3ad3229340c7a8ba1b361ac2c0e).

HeartSaVioR commented:
Never mind. I just had hard reset to commit which follows @cloud-fan approach since it's non-intrusive to Dataset. Old approach is stored in https://github.com/apache/spark/compare/master...HeartSaVioR:SPARK-22000-approach-v1

SparkQA commented:
**[Test build #102692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102692/testReport)** for PR 23854 at commit [`60ca088`](https://github.com/apache/spark/commit/60ca0889f14db3ad3229340c7a8ba1b361ac2c0e).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102692/
Test FAILed.

HeartSaVioR committed:
Fix java lint

SparkQA commented:
**[Test build #102707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102707/testReport)** for PR 23854 at commit [`9eeb391`](https://github.com/apache/spark/commit/9eeb3910889e3809c119cb1b00ba4898d3b0ab0d).

SparkQA commented:
**[Test build #102707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102707/testReport)** for PR 23854 at commit [`9eeb391`](https://github.com/apache/spark/commit/9eeb3910889e3809c119cb1b00ba4898d3b0ab0d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class RecordSpark22000 `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102707/
Test PASSed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -139,7 +139,7 @@ object ScalaReflection extends ScalaReflection {
    * only need to do this for leaf nodes.
    */
-  private def upCastToExpectedType(expr: Expression, expected: DataType,
+  private[spark] def upCastToExpectedType(expr: Expression, expected: DataType,
```
How about moving this function into `object Upcast {}`?

HeartSaVioR replied:
Nice suggestion! Btw, it might be a part of refactor between `deserializerForType` so may not need to have its own object.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -349,10 +349,18 @@ object ScalaReflection extends ScalaReflection {
+          s"""- map key class: "$classNameForKey",
+             | value class: "$classNameForValue"""".stripMargin +:
+          walkedTypePath
+
```
Is this change related to this pr?

HeartSaVioR replied:
Not strictly, but consistent between `ScalaReflection.deserializeFor` and `JavaTypeInference.deserializeFor`. It is marked as TODO and symmetric part is addressed in `JavaTypeInference.deserializeFor` so why not address it to `ScalaReflection.deserializeFor` too? 

If we would like to address this from out of this PR I'll add follow-up PR against this.

maropu replied:
Can you add tests for this code path in `ScalaReflectionSuite`?

HeartSaVioR replied:
Adding a UT just for testing path addition looks overkill. If you're unsure about this chance unless I add a new test for this, I'd rather roll back (adding TODO back) and address this in follow-up PR which may add a new object tracking walkedTypePath for both Java/Scala.

maropu commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanDeserializationSuite.java:
```diff
@@ -115,6 +119,74 @@ public void testBeanWithMapFieldsDeserialization() {
 
+  @Test
+  public void testSpark22000() {
+    List<Row> inputRows = new ArrayList<>();
```
Could you add more tests for upcast failure cases?

HeartSaVioR replied:
Honestly I'm not sure which cases other that this fail without upcast. Please share if you have any, and I'll add them.

maropu replied:
How about this case?
```

scala> :paste
class BeanClass extends Serializable {
  private var id: Int = _
  def getId(): Int = id
  def setId(i: Int): Unit = { id = i }
}
implicit val encoder = org.apache.spark.sql.Encoders.bean(classOf[BeanClass])
Seq("1", "2", "a").toDF("id").as[BeanClass].collect

org.apache.spark.sql.AnalysisException: Cannot up cast `id` from string to int as it may truncate
The type path of the target object is:
- field (class: "int",
name: "id")
- root class: "BeanClass"
You can either add an explicit cast to the input data or choose a higher precision type of the field in the target object;
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveUpCast$.org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveUpCast$$fail(Analyzer.scala:2536)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveUpCast$$anonfun$apply$29$$anonfun$applyOrElse$129.applyOrElse(Analyzer.scala:2552)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveUpCast$$anonfun$apply$29$$anonfun$applyOrElse$129.applyOrElse(Analyzer.scala:2547)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDown$1(TreeNode.scala:258)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:72)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:258)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDown$3(TreeNode.scala:263)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$mapChildren$1(TreeNode.scala:328)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:189)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:326)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:263)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDown$3(TreeNode.scala:263)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$mapChildren$2(TreeNode.scala:345)
```

HeartSaVioR replied:
Ah OK I misread your comment. Looks like you would like to have another test which can't be mitigated with upcast so the query is failing even with upcast. I'll add it. Thanks!

maropu replied:
Thanks!

HeartSaVioR replied:
Addressed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -194,13 +195,31 @@ object JavaTypeInference {
+      expr
+    } else {
+      AssertNotNull(expr, walkedTypePath)
+    }
```
Since this code structure is similar to `ScalaReflection.deserializerForType`, could we share the code between them?

HeartSaVioR replied:
Yes that sounds better, wondering which name of object (maybe) is correct. Would `DeserializerHelper` make sense? I'll try to refactor.

maropu reviewed (commented):
Thanks for your work! The approach itself looks nice to me! I left some comments, so could you check?

HeartSaVioR committed:
Refactor: extract duplicated codes in deserializerFor between ScalaReflection and JavaTypeInference

SparkQA commented:
**[Test build #102736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102736/testReport)** for PR 23854 at commit [`ca124b9`](https://github.com/apache/spark/commit/ca124b92a39d36353dfd56abe5edc23e113377b0).

HeartSaVioR commented:
d4c2060 only changes parameter names from ca124b9 so the build result would be expected to be same.

Btw, while d4c2060 is refactoring on `deserializerFor` between `ScalaReflection` and `JavaTypeInference`, I'm seeing similar needs on `serializerFor`, as well as `walkedTypePath`. Would we want to have them refactored similarly? If then I'll work on that in separate JIRA & PR.

Thanks!

SparkQA commented:
**[Test build #102737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102737/testReport)** for PR 23854 at commit [`d4c2060`](https://github.com/apache/spark/commit/d4c2060d41a2aa48bb0c3f6feeb32c9ca124cf85).

HeartSaVioR committed:
Add new test: upcast doesn't help making Dataset be compatible with Java Bean Encoder

SparkQA commented:
**[Test build #102739 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102739/testReport)** for PR 23854 at commit [`9c040e2`](https://github.com/apache/spark/commit/9c040e2dff215168633bd2e27b3695dbe0a00383).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -212,82 +219,92 @@ object JavaTypeInference {
-              inferDataType(elementType)._1),
-            "array",
-            ObjectType(c))
+        val arrayCls = MapObjects(mapFunction, path, dataType)
```
nit: we can move the `mapFunction` and `arrayCls` into `if (elementNullable)`

HeartSaVioR replied:
Yeah my bad. Moved the code and renamed to `arrayData`.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -212,82 +219,92 @@ object JavaTypeInference {
         val (keyType, valueType) = mapKeyValueType(typeToken)
+        val newTypePath = s"""- map key class: "${keyType.getType.getTypeName}",
+                              | value class: "${valueType.getType.getTypeName}"""" +:
+          walkedTypePath
```
the code style here is very weird, how about
```
val newTypePath = ("- map key class: ..." +
  "value class: ...") +: walkedTypePath
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -300,25 +317,26 @@ object JavaTypeInference {
+          val (dataType, nullable) = inferDataType(fieldType)
+          val newTypePath =
+            s"""- field (class: "${fieldType.getType.getTypeName}",
+               |name: "$fieldName")""".stripMargin +: walkedTypePath
```
ditto

SparkQA commented:
**[Test build #102739 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102739/testReport)** for PR 23854 at commit [`9c040e2`](https://github.com/apache/spark/commit/9c040e2dff215168633bd2e27b3695dbe0a00383).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class RecordSpark22000FailToUpcast `

SparkQA commented:
**[Test build #102736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102736/testReport)** for PR 23854 at commit [`ca124b9`](https://github.com/apache/spark/commit/ca124b92a39d36353dfd56abe5edc23e113377b0).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   * For example, we build an encoder for `case class Data(a: Int, b: String)` and the real type`

SparkQA commented:
**[Test build #102737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102737/testReport)** for PR 23854 at commit [`d4c2060`](https://github.com/apache/spark/commit/d4c2060d41a2aa48bb0c3f6feeb32c9ca124cf85).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   * For example, we build an encoder for `case class Data(a: Int, b: String)` and the real type`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102739/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102737/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102736/
Test FAILed.

HeartSaVioR committed:
Address review comments from cloud-fan

HeartSaVioR committed:
Address weird style from ScalaReflection as well

SparkQA commented:
**[Test build #102751 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102751/testReport)** for PR 23854 at commit [`4d564ab`](https://github.com/apache/spark/commit/4d564abbe39aaa7c61d2dd12b45cd46c557ce889).

SparkQA commented:
**[Test build #102751 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102751/testReport)** for PR 23854 at commit [`4d564ab`](https://github.com/apache/spark/commit/4d564abbe39aaa7c61d2dd12b45cd46c557ce889).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102751/
Test FAILed.

HeartSaVioR commented:
Java style test may be false alarm: failure is not due to violation but inaccessible of dtd link. 
Just a submitted fix for this: #23887

HeartSaVioR commented:
Retest this please.

SparkQA commented:
**[Test build #102762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102762/testReport)** for PR 23854 at commit [`4d564ab`](https://github.com/apache/spark/commit/4d564abbe39aaa7c61d2dd12b45cd46c557ce889).

SparkQA commented:
**[Test build #102762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102762/testReport)** for PR 23854 at commit [`4d564ab`](https://github.com/apache/spark/commit/4d564abbe39aaa7c61d2dd12b45cd46c557ce889).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102762/
Test FAILed.

HeartSaVioR committed:
Fix build failure due to slightly change of walked type path representation

SparkQA commented:
**[Test build #102768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102768/testReport)** for PR 23854 at commit [`e01bfe6`](https://github.com/apache/spark/commit/e01bfe645f5d27e1cae1f703fdd3a76bb168291e).

SparkQA commented:
**[Test build #102768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102768/testReport)** for PR 23854 at commit [`e01bfe6`](https://github.com/apache/spark/commit/e01bfe645f5d27e1cae1f703fdd3a76bb168291e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102768/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -0,0 +1,158 @@
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.types._
+
+private[spark] object DeserializerBuildHelper {
```
nit: everything in catalyst is private, we don't need the `private[spark]`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -211,73 +167,53 @@ object ScalaReflection extends ScalaReflection {
       case t if t <:< localTypeOf[java.math.BigDecimal] =>
-        Invoke(path, "toJavaBigDecimal", ObjectType(classOf[java.math.BigDecimal]),
-          returnNullable = false)
+        createDeserializerForJavaBigDecimal(path, returnNullable = false)
```
do you mind taking some time to investigate it? Why the java side uses `returnNullable = true` but scala side uses `returnNullable = false`?

maropu replied:
`createDeserializerForJavaBigInteger`, too.

HeartSaVioR replied:
Would we mind if we file a new issue which tracks the effort to investigate it? IMHO it looks like beyond this PR. and I'm seeing multiple PRs which keep adding requirements and suddenly reviewers lost interest so fail to reach conclusion.

cloud-fan replied:
yea fine to do it in followup.

HeartSaVioR replied:
Filed https://issues.apache.org/jira/browse/SPARK-26991

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  /** Returns the current path with a field at ordinal extracted. */
+  def addToPathOrdinal(
```
This function is only used in `ScalaReflection`? If so, how about moving this into there and make it private?

HeartSaVioR replied:
It might be weird if we split out similar code (basically this only adds index handling on `addToPath`) to multiple places, but no strong opinion. WDYT? If you still feel better to move this to ScalaReflection, please let me know so that I can move it.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -0,0 +1,158 @@
+    upCastToExpectedType(newPath, dataType, walkedTypePath)
+  }
+
+  def expressionWithNullSafety(
```
private?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -212,82 +219,90 @@ object JavaTypeInference {
+            case c if c == java.lang.Short.TYPE => "toShortArray"
+            case c if c == java.lang.Byte.TYPE => "toByteArray"
+            case c if c == java.lang.Boolean.TYPE => "toBooleanArray"
+            case other => throw new IllegalStateException("expect primitive array element type " +
```
Why we don't allow non-nullable & non-primitive case?

HeartSaVioR replied:
Ah my bad. It was for sync between ScalaReflection and JavaTypeInference and I realized it is not exactly same. I'll roll back the change.

maropu replied:
Thanks!

maropu replied:
btw, we don't have any test for this code path? (it seems the latest Jenkins test run passed though...)

HeartSaVioR replied:
I guess it is very hard to cover all cases, especially in Java test we rely on defining actual bean class to verify. It would be nice to cover more cases if we could define (column type, field type) matrix and create classes. Maybe another JIRA issue for this?

maropu replied:
WDYT, @cloud-fan ?

HeartSaVioR replied:
@maropu 
Sorry to make a confusion. I wasn't wrong so the change was unnecessary, so I'm rolling back the change. To answer the origin comment, non-nullable & non-primitive case => not possible according to the implementation of `inferDataType`. `nullable = true` for only primitive case.

cloud-fan replied:
> non-nullable & non-primitive case => not possible

I agree with this, and I think the original code style is better, i.e.
```
val primitiveMethod = ...
val primitiveMethod.map {
  ...
}.getOrElse {
  ...
}
```

We can change the scala side to follow this.

HeartSaVioR replied:
OK. Would we keep `mapFunction` to deal with upcast? I guess `primitiveMethod.map.getOrElse` gets rid of necessary of `if (elementNullable)`, but just confirm again whether we would want to go back to origin code which upcast was not there, or just take previous code style.

cloud-fan replied:
we definitely need to do upcast, just take the code style.

HeartSaVioR replied:
Thanks for clarification! Just addressed.

HeartSaVioR committed:
Address review comments from cloud-fan & maropu

SparkQA commented:
**[Test build #102782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102782/testReport)** for PR 23854 at commit [`8cbad26`](https://github.com/apache/spark/commit/8cbad26298bd57c852e3847ad21b1fd64df5cfb2).

SparkQA commented:
**[Test build #102782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102782/testReport)** for PR 23854 at commit [`8cbad26`](https://github.com/apache/spark/commit/8cbad26298bd57c852e3847ad21b1fd64df5cfb2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102782/
Test FAILed.

HeartSaVioR commented:
retest this please

SparkQA commented:
**[Test build #102785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102785/testReport)** for PR 23854 at commit [`8cbad26`](https://github.com/apache/spark/commit/8cbad26298bd57c852e3847ad21b1fd64df5cfb2).

SparkQA commented:
**[Test build #102785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102785/testReport)** for PR 23854 at commit [`8cbad26`](https://github.com/apache/spark/commit/8cbad26298bd57c852e3847ad21b1fd64df5cfb2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102785/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -212,82 +219,99 @@ object JavaTypeInference {
+          primitiveMethod.map { method =>
+            Invoke(path, method, ObjectType(c))
+          }.getOrElse {
+            Invoke(
```
what's the difference between this branch and `if (elementNullable)`? Shall we follow the previous code? e.g.
```
val primitiveMethod = ...
primitiveMethod.map(...).getOrElse...
```

HeartSaVioR replied:
Hmm... sorry I think I was a bit confused. After looking into it again, the content before this commit seems correct and also in sync with ScalaReflection. I'll revert it back.
Btw, the difference between previous code is that new code handles upcast which ScalaReflection also handles it.

HeartSaVioR committed:
Revert back wrong change and sync again with ScalaReflection implementation

HeartSaVioR committed:
Apply preferred code style to both JavaTypeInference and ScalaReflection

SparkQA commented:
**[Test build #102796 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102796/testReport)** for PR 23854 at commit [`24a1b19`](https://github.com/apache/spark/commit/24a1b195bad49e05353b80c85d30e32e8c898a52).

SparkQA commented:
**[Test build #102796 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102796/testReport)** for PR 23854 at commit [`24a1b19`](https://github.com/apache/spark/commit/24a1b195bad49e05353b80c85d30e32e8c898a52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102796/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

HeartSaVioR commented:
Thanks all for reviewing, providing nice approaches, and finally merging!
